core/Date.php

```diff
@@ -611,6 +612,38 @@ public function subYear($n)
         return new Date($result, $this->timezone);
     }
 
+    protected static $use12HourClock;
+
+    protected function getTimeFormat()
+    {
+        if (is_null(self::$use12HourClock)) {
+            self::$use12HourClock = LanguagesManager::uses12HourClockForCurrentUser();
```

Comment:
We should avoid using DI to store setting info, since settings can change in the middle of a request while DI is meant to be static for the whole request. Ie, if we change this setting in the middle of a request, there's no way to re-inject the objects that have already been created.

Instead, what I'd do is:
1) Create a DateTimeFormatProvider service class in core to return individual formats. This class would provide default translations w/o going through the translator.
2) Create a new DateTimeFormatProvider in the Intl/LanguagesManager plugin (whichever provides the best behavior when one of the plugins is not loaded) which uses the Translator to provide the formats.
3) Override the DateTimeFormatProvider key in the plugin's config.php to use the plugin's DateTimeFormatProvider.
2) Use the DateTimeFormatProvider here.

I'd probably also move the date/time formatting to MetricsFormatter (& move MetricsFormatter to DI as well), however that's not really required at the moment. And I'll have to do something like that to get rid of StaticContainer eventually, anyway.